apps/web/src/app/(networks)/(non-evm)/stellar/swap/lib/swap-get-route/use-pool-graph.ts

```diff
@@ -0,0 +1,228 @@
+} from '~stellar/_common/lib/soroban/client'
+import { isAddressLower } from '~stellar/_common/lib/soroban/constants'
+import { contractAddresses } from '~stellar/_common/lib/soroban/contracts'
+import type { Vertex } from './types'
+
+type GetPoolsByTokenPairsBatchedParams = Array<{
+  token_a: string
+  token_b: string
+  fee: number
+}>
+
+type PoolData = {
+  token0: string
+  token1: string
+  tick: number
+  sqrtPriceX96: bigint
+  liquidity: bigint
+  poolAddress: string
+  fee: number
+}
+
+const getPoolsByTokenPairsBatched = async (
+  params: GetPoolsByTokenPairsBatchedParams,
+): Promise<PoolData[]> => {
+  const BATCH_SIZE = 3
+  const batches: GetPoolsByTokenPairsBatchedParams[] = []
+  const poolLensContractClient = getPoolLensContractClient({
+    contractId: contractAddresses.POOL_LENS,
+  })
+
+  for (let i = 0; i < params.length; i += BATCH_SIZE) {
+    batches.push(params.slice(i, i + BATCH_SIZE))
+  }
+
+  const results: PoolData[][] = await Promise.all(
+    batches.map(async (batch) => {
+      const batchPoolsResult = await poolLensContractClient.get_pools_by_pairs({
+        pairs: batch,
+      })
+      const batchedPoolsInfo: PoolData[] = []
+      for (const poolResult of batchPoolsResult.result) {
+        if (poolResult.result.tag === 'NotFound') {
+          continue
+        }
+        const [poolData] = poolResult.result.values
+        batchedPoolsInfo.push({
+          token0: poolData.token0,
+          token1: poolData.token1,
+          tick: poolData.tick,
+          sqrtPriceX96: poolData.sqrt_price_x96,
+          liquidity: poolData.liquidity,
+          poolAddress: poolResult.pool,
+          fee: poolData.fee,
+        })
+      }
+      return batchedPoolsInfo
+    }),
+  )
+
+  return results.flat()
+}
+
+/**
+ * Build a graph of available pools for routing
+ *
+ * Returns:
+ * - vertices: Map of pool vertices (tokenA|||tokenB -> pool data)
+ * - tokenGraph: Map of token adjacency list (token -> connected tokens)
+ */
+export function usePoolGraph() {
+  return useQuery({
+    queryKey: ['stellar', 'pool-graph'],
+    queryFn: async () => {
+      // Store arrays of vertices per token pair to handle multiple fee tiers
+      const vertices = new Map<string, Vertex[]>()
+      const tokenGraph = new Map<string, string[]>()
+
+      try {
+        // Get factory client to discover pools
+        const factoryClient = getFactoryContractClient({
+          contractId: contractAddresses.FACTORY,
+        })
+
+        if (!factoryClient) {
+          console.error('Failed to create factory client')
+          return { vertices, tokenGraph }
+        }
+
+        // For performance, we'll query a subset of known token pairs
+        // In production, you'd want to:
+        // 1. Query all pools from the factory
+        // 2. Cache results
+        // 3. Only query active/liquid pools
+        const knownTokens = staticTokens.map((token) => token.contract)
+
+        const feeTiers = await getFees()
```

Comment:
See previous comment about this. Don't know how this could still return a route. Unless pools were created with the wrong fee too? Although the factory should stop it from happening

Reply:
See https://github.com/sushi-labs/sushiswap/pull/2043#discussion_r2623645288

Reply:
Right, gotcha, expected to see FEE_TIERS used here. fine with whatever

Reply:
This has been refactored to make it clearer (kept this logic separate from `TICK_SPACINGS`.